test/zipkin/util.go

```diff
@@ -0,0 +1,135 @@
+/*
+Copyright 2018 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+//zipkin_util has constants and helper methods useful for zipkin tracing support.
+
+package zipkin
+
+import (
+	"errors"
+	"fmt"
+	"net"
+	"os"
+	"os/exec"
+	"strconv"
+
+	"github.com/knative/pkg/test/logging"
+	"go.opencensus.io/trace"
+	"k8s.io/client-go/kubernetes"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+
+	// ZipkinPortForwardPid environment variable name to store Process ID of the process started for zipkin port-foward.
```

Comment:
I only see this being used in the setup/cleanup methods. Why does it need to be an environment variable? Why not a global variable?

Reply:
The thinking here was that since this process would be outside the Go process (test) it would be better to store the information about it in an environment variable rather than a variable inside the Go process. Especially if the test is somehow cut short for any reason and Cleanup(..) method doesn't get a chance to run. Having it in an environment variable would be more easy to find(still would require a manual kill).

Reply:
I don't see where you're spawning a child process in order to use that, but maybe I'm missing something?

Reply:
This statement : err = portForwardCmd.Start() starts the child process which is tracked in the environment variable.  Start() returns by creating the child process whose Process ID is stored in Cmd.Process.Pid

Reply:
Yes, I'm aware of all that. But as I said, you're setting the variable later, so `kubectl` is not using it. I still don't see where and why you're using an environment variable here instead of a simple global variable.

Reply:
Sorry was just trying to answer your question "I don't see where you're spawning a child process". The "kubectl" itself is the child process as the port-forward command doesn't terminate till the process that started itself is killed. So what I was doing is storing the process id of the kubectl command in the environment variable and then using that in the cleanup method to kill it. 

The rationale behind of using an environment variable instead of global variable was explained in my earlier comment - The test could be killed before cleanup method gets a chance to run. In such cases having it in env variable would be easier then a global variable.

I have updated to use global variable. But also wanted to make sure I explain the rationale behind.

Reply:
Unless I'm missing something here, I don't see how the environment variable is helping you here. If the test can be killed before the cleanup phase, why the cleanup phase survives? And if it survives, it should be able to read the PID from a local resource. You're not using the env var in a subprocess, you're just using it as a storage area. I would understand that in a sibling UNIX process context, which doesn't seem to be the case here. Correct me if I'm wrong.

Reply:
Yes, if the test is killed before the cleanup phase, the cleanup phase would not survive. In such a scenario it would require a manual kill of the port-forward process. This might come up when developer is testing in local box and cut short the test by Crtl+C. My rationale was in such scenario it would be easier to find the Process to kill if the process-id is stored in an environment variable. Hence in my first comment I mentioned : "Having it in an environment variable would be more easy to find(still would require a manual kill)."

Reply:
You won't be able to inspect such environment variable in the given scenario because your process will be in different shells. In this case, logging the PID is the best option.

Reply:
Ah got it. I have added a log statement.